### DISCUSSION TITLE
Publish message with confirm but still message is lost without error?

### DISCUSSION THREAD

**Discussion by rabbitLove520** [rabbitmq/rabbitmq-dotnet-client#1284] [General]
**RabbitMQ.Client version 6.4.0**
**dotnet 7.0**

my send code:
```csharp
public void SendToQueue(string data, string? routingKey)
{
    try 
    {
        channelPublish.ConfirmSelect(); // confirm mode
        var properties = channelPublish.CreateBasicProperties();
        properties.DeliveryMode = 2;
        properties.Persistent = true;
        properties.ContentType = "application/json";
        properties.ContentEncoding = "UTF-8";
        var newBytes = Encoding.UTF8.GetBytes(data);

        channelPublish.BasicPublish(exchangeName,
            routingKey: routingKey, 
            basicProperties: properties, 
            body: newBytes,
            mandatory: true
        );
        channelPublish.WaitForConfirmsOrDie();
    }
    catch (Exception ex)
    {
        _logger.LogError("Send to queue failed: " + ex.ToString());
    }
}
```
I found sometimes message is lost, RabbitMQ server with some **strange** log:
```
2022-12-16 14:23:37.311 [warning] <0.5713.0> closing AMQP connection <0.5713.0> (183.157.66.178:57208 -> 10.99.55.2:5672, vhost: '/', user: 'happy'):
client unexpectedly closed TCP connection
2022-12-16 14:23:37.312 [info] <0.5874.0> Closing all channels from connection '183.157.66.178:57208 -> 10.99.55.2:5672' because it has been **closed**
```
I shoud got error when connection is lost, however I got nothing.

**Comment by michaelklishin** [comment]
Those "strange logs" [tell you something important](https://www.rabbitmq.com/logging.html#logged-events): your client closed or lost TCP connection. Your code does not [register any shutdown handlers](https://www.rabbitmq.com/dotnet-api-guide.html#consuming) so you only can find out about this from RabbitMQ logs.

There are multiple separate failure scenarios that you assume are one (or all handled by publisher confirms):

 * Connection loss. There is automatic recovery support but publisher confirms simply allow you to notice that some data hasn't been acknowledged and therefore *must* be resent by your application.
 * Channel closure, e.g. [due to an exception](https://www.rabbitmq.com/channels.html#error-handling). There is no automatic recover by design but as far as publisher confirms go, the idea is the same.
 * When a target queue could not replicate the message. This is where you'd get [negative publisher confirms](https://www.rabbitmq.com/publishers.html#data-safety).
 * Unroutable messages. You would not get negative publisher confirms but you can get a [returned message](https://www.rabbitmq.com/dotnet-api-guide.html#basic-return) if you publisher as `mandatory` and the [unroutable message metric](https://www.rabbitmq.com/publishers.html#metrics) will be `> 0`.

In other words, you application use publisher confirms to find out when they must resent one or a group of messages. There are other failures scenarios that publisher confirms do not cover,
and cannot cover (e.g. the messages could not even get to RabbitMQ to in the first place).

**Reply by rabbitLove520** [reply]
My message must be routable because routingKey is same, just wonder why I can't get any exception on operating the channel when Connection is lost.

**Reply by rabbitLove520** [reply]
I have registered `ConnectionShutdown` event:
  ```
connection.ConnectionShutdown += (o, a) => {
      Console.WriteLine("Connection Shutdown");
  };
```
but I couldn't find any connection shutdown log while RabbitMQ server showed `client unexpectedly closed TCP connection`😭😭

**Comment by lukebakken** [comment]
Your code should have caught this exception:

```
new IOException("timed out waiting for acks")
```

Did it?

Please avoid using `WaitForConfirmsOrDie`. See this tutorial for a much better way to track publisher confirmations:

https://www.rabbitmq.com/tutorials/tutorial-seven-dotnet.html

**Reply by rabbitLove520** [reply]
I didn't catch any exception..

**Reply by OlegStotsky** [reply]
I'm experiencing the same issue in the go client. I receive confirmation for the message however it is never delivered to the client.

**Reply by lukebakken** [reply]
@OlegStotsky - your comment here is nearly useless, here's why:
* This is the incorrect repository to make the comment, and you're commenting in an old discussion.
* You have provided absolutely no details about your issue. No code, no logs, no proof that the issue lies outside of your own code.

If you would like assistance, start a discussion here - https://github.com/rabbitmq/amqp091-go/discussions/
You _MUST_ include the following information or we will not assist you:
* RabbitMQ, Erlang and operating system version
* Code that we can clone, compile, and run to see the issue happen
* Any other relevant details - see this guide: https://github.com/rabbitmq/support-tools/blob/main/docs/Reporting_RabbitMQ_Issues.md

**Reply by michaelklishin** [reply]
> I'm experiencing the same issue in the go client. I receive confirmation for the message however it is never delivered to the client.

I have a feeling that you expect a publisher confirm to be sent when a consumer gets a delivery. 

That's not how publisher confirms work: they are not aware of consumers. When exactly they will be sent is covered in [the relevant doc guide](https://www.rabbitmq.com/docs/confirms#publisher-confirms). They cover communication between the publisher and the target queue (the exact semantics will vary from non mirrored classic queues and replicated quorum queues)

Likewise, consumer acknowledgement are entirely unaware of publishers and cover consumer delivery (and processing).

**Reply by lukebakken** [reply]
> I have a feeling that you expect a publisher confirm to be sent when a consumer gets a delivery

Ah, good thinking @michaelklishin

**Comment by michaelklishin** [comment]
We will not allow single sentence problem definitions and discussions of the Go client here. In fact, spamming existing discussions in random repos repeatedly is a good way to get an org-wide ban.

Those who are experiencing "the same issue" should use an appropriate client library repo Discussions or the RabbitMQ server repo Discussions, and provide an executable way you can reproduce the behaviour.

We do not guess in this community and the burden of proof/demonstration is on the reporter, not the core team.
